File: swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def traverse(self,prev, curr):
        if not curr or not curr.next:
            return
        
        nxt_pair = curr.next.next
        nxt = curr.next
        
        nxt.next = curr
        curr.next = nxt_pair
        prev.next = nxt
        
        
        prev = curr
        curr = nxt_pair
        
        
        

        

        self.traverse(prev, curr)
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        sentinel = ListNode(val=None, next=head)
        curr = sentinel.next
        
        self.traverse(sentinel, curr)
        return sentinel.next
        

File: 2020-algorithms-studying/LinkedLists/SwapNodesInPairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    if(head === null) return head
    if(head.next === null) return head
    /**
    The first is you go througbh and swap the values

    the second is you go through and swap the node
    - Two pointers
    
    
    if odd don't swap last node
    
    
    **/
    let curr = head
        while(curr){

    if(curr.next){
    let first = curr.val
    let second = curr.next.val
    curr.next.val = first
    curr.val = second
    }
    if(curr.next?.next !== null){
    curr = curr.next?.next

    }else{
        curr = curr.next
    }
    }
    
    return head
    
    
};